### PR TITLE
Inliner: ensure all statements have an inline context

### PR DESCRIPTION
**Pull request by AndyAyersMS**
We would previously allow some statements inserted by the inliner
"between" the caller and callee to have a null inline context. This
lack of context could lead to incomplete or misleading inline trees
if the inserted code included calls. In particular some failed inlines
might not show up in the trees.

Set the context for these inserted statements to be the context for the
call being inlined (thus they logically "belong" to the caller).

There are quite a few variables typed as `GenTreePtr` that must point
at statements -- for instance `fgMorphStmt`. Retype these as `GenTreeStmt*`.
Cleanup dereferences which no longer require a `gtStmt`.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
Left as two commits for now. No diffs in CoreCLR testing.

@dotnet/jit-contrib PTAL

**Comment by AndyAyersMS** [comment]
@dotnet-bot retest  Linux ARM Emulator Cross Debug Build

**Comment by AndyAyersMS** [comment]
Updated past merge conflict

**Comment by AndyAyersMS** [comment]
@dotnet/jit-contrib anyone?

File: src/jit/flowgraph.cpp [hunk 1]
```diff
@@ -22002,32 +22018,43 @@ void Compiler::fgInsertInlineeBlocks(InlineInfo* pInlineInfo)
...
+    IL_OFFSETX callILOffset = callStmt->gtStmtILoffsx;
+    GenTreeStmt* postStmt  = callStmt->gtNextStmt;
     GenTreePtr afterStmt = callStmt; // afterStmt is the place where the new statements should be inserted after.
-    GenTreePtr newStmt;
-
+    GenTreePtr newStmt = nullptr;
     GenTreePtr call = inlineInfo->iciCall;
+
     noway_assert(call->gtOper == GT_CALL);
+    noway_assert(callStmt->gtOper == GT_STMT);
```
Comment (JosephTremoulet) [inline src/jit/flowgraph.cpp]: Seems redundant now that you've re-typed `callStmt`

File: src/jit/flowgraph.cpp [hunk 2]
```diff
@@ -22002,32 +22018,43 @@ void Compiler::fgInsertInlineeBlocks(InlineInfo* pInlineInfo)
...
+// Notes:
+//    Statements prepended may include the following:
+//    * This pointer null check
+//    * Class initialization
+//    * Zeroing of must-init locals in the callee
+//    * Passing of call arguments via temps
+//
+//    Newly added statements are placed just after the original call
+//    and are are given the same inline context as the call any calls
+//    added here will appear have been part of the immediate caller.
```
Comment (JosephTremoulet) [inline src/jit/flowgraph.cpp]: typo: should read "will appear **to** have been"

Comment (AndyAyersMS) [inline src/jit/flowgraph.cpp]: Thanks, fixed.

**Pull request merged**